config.go

```diff
@@ -7,9 +7,89 @@ import (
 	"time"
 )
 
+// These are the versions of the protocol (which includes RPC messages as
+// well as Raft-specific log entries) that this server can _understand_. Use
+// the ProtocolVersion member of the Config object to control the version of
+// the protocol to use when _speaking_ to other servers. This is not currently
+// written into snapshots so they are unversioned. Note that depending on the
```

Comment:
One thing we need to guarantee is that if you don't understand something in your log or snapshot, you crash. It's not cool to skip it like we're doing now.

This could be a server was running new code, wrote out a log entry/snapshot with important new information, then somehow got downgraded.
 * For log entries, we can probably agree to do this using new log entry types. But we're missing the check in NewRaft and Recover right now.
 * For snapshots, we don't have a way to know.

Otherwise, this could be that we've deleted the code that reads in old log entries/snapshot formats. Someday we need to remove that cruft. How can we guarantee that no log entry or snapshot is so old? I think the solution has to be of the form: write out a new snapshot in a newer format before you upgrade to code that stops knowing how to read older log entry/snapshot formats. How do we arrange that?

BTW, I think all this might lead us to having multiple versions covering different things, at least internally. You can't read in snapshots with newer versions than what you support, but the snapshot format won't change on every upgrade. So I think it needs its own class of version number.